fair/inverse.py

```diff
@@ -0,0 +1,135 @@
+        rt            : sensitivity of time-integrated airborne fraction to
+                        temperature (yr/K)
+        iirf_max      : maximum value of time-integrated airborne fraction (yr)
+        time_scale_sf0: initial guess of alpha scaling factor
+        a             : partition coefficient of carbon boxes
+        tau           : present-day decay time constants of CO2 (yr)
+        iirf_h        : time horizon for time-integrated airborne fraction (yr)
+        carbon_boxes0 : carbon stored in each atmospheric reservoir at timestep
+                        t-1 (GtC)
+        c_pi          : pre-industrial concentration of CO2, ppmv
+        c0            : concentration of CO2 in timestep t-1, ppmv
+        e0            : emissions of CO2 in timestep t, GtC
+
+    Outputs:
+        e1            : emissions of CO2 in timestep t, GtC
+        c_acc1        : cumulative airborne carbon anomaly (GtC) since
+                        pre-industrial, timestep t
+        carbon_boxes1 : carbon stored in each atmospheric reservoir at timestep
+                        t (GtC)
+        time_scale_sf : scale factor for CO2 decay constants
+    """
+     
+    iirf = iirf_simple(c_acc0, temp, r0, rc, rt, iirf_max)
+    time_scale_sf = root(iirf_interp, time_scale_sf,
+      args=(a, tau, iirf_h, iirf))['x']
+    tau_new = tau * time_scale_sf
+    e1 = root(infer_emissions, e0, args=(c1, carbon_boxes0, tau_new, a, c_pi))['x']
+    c_acc1 = c_acc0 + 0.5*(e1 + e0) - (c1 - c0)*ppm_gtc
+    carbon_boxes1 = carbon_boxes0*np.exp(-1.0/tau_new) + a*e1 / ppm_gtc
+    return e1, c_acc1, carbon_boxes1, time_scale_sf
+
+    
+def inverse_fair_scm(
+# Longer term goal: one calling interface only - appropriate function
+# determined from call
+    C             = None,
+    other_rf      = 0.0,
+    q             = thermal.q,
+    tcrecs        = thermal.tcrecs,
+    d             = thermal.d,
+    F2x           = thermal.f2x,
+    tcr_dbl       = thermal.tcr_dbl,
+    a             = carbon.a,
+    tau           = carbon.tau,
+    r0            = carbon.r0,
+    rc            = carbon.rc,
+    rt            = carbon.rt,
+    iirf_max      = carbon.iirf_max,
+    iirf_h        = carbon.iirf_h,
+    C_pi          = 278.,
+    time_scale_sf = 0.16
+    ):
+    
+    # Error checking and validation goes here...
+
+    # Dimensions
+    nt = len(C)
+    carbon_boxes_shape = (nt, a.shape[0])
+    thermal_boxes_shape = (nt, d.shape[0])
+    
+    # Thermal response
+    q         = calculate_q(tcrecs, d, F2x, tcr_dbl, nt)
+    
+    # Allocate intermediate and output arrays
+    C_acc     = np.zeros(nt)
+    R_i       = np.zeros(carbon_boxes_shape)
+    emissions = np.zeros(nt)
+    T_j       = np.zeros(thermal_boxes_shape)
+    F         = np.zeros(nt)
+    
+    if np.isscalar(other_rf):
+        other_rf = other_rf * np.ones(nt)
+    
+    # First timestep
+    emissions[0] = root(infer_emissions, 0., args=(C[0], R_i[0],
```

Comment:
If you want a speed improvement, you can invert a forward-difference approximation by hand. Maybe one for future rather than now.

Reply:
Yeah, I can't quite get this to check out - I feel there's an obvious solution but I can't find it.

Reply:
leave it for 2.0